content/post/graphql-with-go-and-postgresql.md

```diff
@@ -0,0 +1,377 @@
+
+```go
+var CommentType = graphql.NewObject(graphql.ObjectConfig{
+	Name: "Comment",
+	Fields: graphql.Fields{
+		"id": &graphql.Field{
+			Type: graphql.NewNonNull(graphql.ID),
+			Resolve: func(p graphql.ResolveParams) (interface{}, error) {
+				if comment, ok := p.Source.(*Comment); ok == true {
+					return comment.ID, nil
+				}
+				return nil, nil
+			},
+		},
+		"title": &graphql.Field{
+			Type: graphql.NewNonNull(graphql.String),
+			Resolve: func(p graphql.ResolveParams) (interface{}, error) {
+				if comment, ok := p.Source.(*Comment); ok == true {
+					return comment.Title, nil
+				}
+				return nil, nil
+			},
+		},
+		"body": &graphql.Field{
+			Type: graphql.NewNonNull(graphql.ID),
+			Resolve: func(p graphql.ResolveParams) (interface{}, error) {
+				if comment, ok := p.Source.(*Comment); ok == true {
+					return comment.Body, nil
+				}
+				return nil, nil
+			},
+		},
+	},
+})
+func init() {
+	CommentType.AddFieldConfig("user", &graphql.Field{
+		Type: UserType,
+		Resolve: func(p graphql.ResolveParams) (interface{}, error) {
+			if comment, ok := p.Source.(*Comment); ok == true {
+				return GetUserByID(comment.UserID)
+			}
+			return nil, nil
+		},
+	})
+	CommentType.AddFieldConfig("post", &graphql.Field{
+		Type: PostType,
+		Args: graphql.FieldConfigArgument{
+			"id": &graphql.ArgumentConfig{
+				Description: "Post ID",
+				Type:        graphql.NewNonNull(graphql.ID),
+			},
+		},
+		Resolve: func(p graphql.ResolveParams) (interface{}, error) {
+			i := p.Args["id"].(string)
+			id, err := strconv.Atoi(i)
+			if err != nil {
+				return nil, err
+			}
+			return GetPostByID(id)
+		},
+	})
+}
+```
+_Listing 4_
+
+Just like in the schema in _Listing 1_, the `Comment` type is a structure with three attributes defined statically; `id`, `title` and `body`. Two other attributes `user` and `comment` are defined dynamically to avoid circular dependencies.
+
+Go does not lend itself well to this kind of dynamic modelling, there is little type checking support, most of the variables in the code are of type `interface{}` and need to be type asserted. `CommentType` itself is a varible of type `graphql.Object` and it's attributes are of type `graphql.Field`. So, there's not direct translation between the GraphQL DSL and the data structures used in go.
+
+The `resolve` function for each field exposes the `Source` parameter which is a data type vertex representing the previous node in the traversal. All the attributes of a `Comment` have, as source, the current `CommentType` vertex. Retrieving the `id`, `title` and `body` are straightforward attribute accessors, while retrieving the `user` and the `post` require graph traversals, and thus database queries. The SQL queries are left out of this document because of their simplicity, but they are available in github repository listed in the references section.
```

Comment:
- Retrieving the ... **is**
- **requires** graph
- **the** github repository ... _References_